### PR TITLE
Add simple `cloned()` and `inspect()` combinators

### PR DESCRIPTION
**Pull request by cuviper**
These work like `Iterator::cloned()` and `Iterator::inspect()`.

### CODE REVIEW

**Comment by cuviper** [comment]
Added the weight, but held off on trying `Cloned` for now.  Let me know if you want it after all.

I'm also thinking of adding `fn inspect`, but that really will need a new type AFAICT.

**Comment by nikomatsakis** [comment]
@cuviper well, I have no problem with implementing `Cloned` as its own specialized iterator type, but I'm also happy to stick with this version for now and implement that later. Another option that occurred to me is that we might add a "layer of indirection" to make this easier in the future.

Specifically, instead of having the "map" iterators take a `MAP_OP: Fn(In) -> Out` directly, they could take a `MAP_OP: MapFn<In, Out>` and we could have wrapper type `ClosureMapFn(F)` that implements MapFn<In,Out>` for any `F: Fn(In) -> Out`. The reason for this indirection is that then we could have a type `struct Cloner` that implements `MapFn<&'a T, T>` for all T: Clone, basically.

**Comment by nikomatsakis** [comment]
I think I like this `MapOp` idea. I suspect there will be other cases of convenience style functions that can be implemented on top of `Map`.

**Comment by nikomatsakis** [comment]
OK, so this branch seems like a good first step, so I'm happy to rebase/merge it and then try to do this `MapOp` idea I had as a follow-up -- or just do it all at once. Let me know. (But the branch needs a rebase anyhow.)

**Comment by cuviper** [comment]
I'd like to try your idea, if you're not already working on it.

**Comment by nikomatsakis** [comment]
No I am not likely to have much time for coding this next few days

**Comment by cuviper** [comment]
OK, done -- let me know what you think of this...

File: src/par_iter/map.rs [hunk 1]
```diff
@@ -14,12 +56,11 @@ impl<M, MAP_OP> Map<M, MAP_OP> {
     }
 }
 
-impl<M, MAP_OP, R> ParallelIterator for Map<M, MAP_OP>
```
Comment (nikomatsakis) [inline src/par_iter/map.rs]: Oh, nice that we can drop the R parameter. That always bugged me.

File: src/par_iter/map.rs [hunk 2]
```diff
@@ -112,16 +148,34 @@ impl<'m, P, MAP_OP, RET> Producer for MapProducer<'m, P, MAP_OP>
...
-    type IntoIter = iter::Map<P::IntoIter, &'m MAP_OP>;
+    type Item = MAP_OP::Output;
+    type IntoIter = MapIter<'m, P::IntoIter, MAP_OP>;
 
     fn into_iter(self) -> Self::IntoIter {
-        self.base.into_iter().map(self.map_op)
+        MapIter {
+            base: self.base.into_iter(),
+            map_op: self.map_op,
+        }
```
Comment (nikomatsakis) [inline src/par_iter/map.rs]: I guess having to write our own version of `map` is the casualty here, huh? Seems like a good trade-off though.

**Comment by nikomatsakis** [comment]
Looks great!

**Pull request merged**